### ISSUE TITLE
Long lived connections stale/become unusable

### ISSUE THREAD

**Issue by ospfranco** [dooboolab/react-native-iap#1562]
### Version of react-native-iap
7.5.0

### Version of react-native
0.65.1

### Platforms you faced the error (IOS or Android or both?)
Android

Hi, we migrated from 6.2.4 to 7.5.0 a month ago, from that moment our error rates started sky-rocketing, we have been trying to debug the issues for some days now, but we couldn't find any issue with our code but there is nothing wrong, however, we did notice somethings:

- We initialize the connection once on app start, and we let the object alive through the app's lifecycle
- On the 6.3.0 release, the logic changed to keep only one client alive (https://github.com/dooboolab/react-native-iap/commit/1409b8f13cb4d3eb60b4c5eac17bab8edaf3cd7d)
- We think a combination of us keeping the billingClient alive for a long period of time + now the handling of a single client instance, causes the client at some point to disconnect and never to reconnect again, causing elevated error rates

Any ideas if this is correct or if something else could be going on?

**Comment by jeremyhalin** [comment]
Hi @ospfranco 

I face the same problem as you (I mentioned my issue that is, I think, related).
What did you do to solve this? Did you downgrde to `^6.0`?

I'm using hooks, do you?

**Comment by ospfranco** [comment]
No, I'm using redux sagas, I added a new method (android only) isReady, if that returns false, you can kill and re-initialize the connection

**Comment by jeremyhalin** [comment]
>No, I'm using redux sagas, I added a new method (android only) isReady, if that returns false, you can kill and re-initialize the connection

Ok thanks, I don't quite know where to add the call of `isReady` in my situation since I'm using hooks.

>you can kill and re-initialize the connection

How do I do that?

In my root container
```javascript
const {
  currentPurchase,
  finishTransaction,
} = useIAP();

useEffect(() => {
  const checkCurrentPurchase = async (purchase): Promise<void> => {
    if (purchase) {
      const receipt = purchase.transactionReceipt;
      if (receipt)
        try {
          await finishTransaction(purchase);
          await dispatch(Purchase.action(purchase))
          // display something
          showMessage({
            message: t('shop.successMessage.message'),
            description: t('shop.successMessage.description'),
            type: 'success',
          })
        } catch (ackErr) {
          console.warn('ackErr', ackErr);
          showMessage({
            message: ackErr.message,
            type: 'danger'
          })
        }
      }
  };

  checkCurrentPurchase(currentPurchase);
}, [currentPurchase, finishTransaction]);
```

And in my Shop container
```javascript
  // ...
  const {
    products,
    getProducts,
    requestPurchase
  } = useIAP()

  // ...

<ShopCardItem
    key={item.productId}
    onPress={() => requestPurchase(item.productId)}
    title={title}
    description={item.description}
    highlight={'🚫'}
    price={item.localizedPrice}
    color={Colors.gold}
/>
```

Do you know if I have to call `init` method with hooks?

**Comment by ospfranco** [comment]
The method is not intended/integrated with hooks, it's meant for manual handling of the client

Cannot help you with your issue

**Comment by mifi** [comment]
Do you know if purchaseUpdatedListener and purchaseErrorListener needs to be re-initialized when endConnection and initConnection is called again?

**Comment by ospfranco** [comment]
Don't think so, the callbacks are handled on the iap package and not the iap instance, shouldn't be too difficult to test it yourself though

**Comment by mifi** [comment]
True! I'll have to do some testing. I just wondered whether it was something you had already thought about in your implementation or if it isn't needed.

**Comment by ospfranco** [comment]
Going to close this issue for now, but here is some info for those looking for some answers:

- Our (temporary) solution was to revert to ~6.x.x version, because that version internally re-initializates the internal iap object whenever some of the methods are called (which causes memory leaks but at least is guaranteed to work)
- The next step is to upgrade again and use the newly exposed method (`isReady`) to check if the instance is alive before doing any operation, if not alive then `initConnection` needs to be called again
- `isReady` is meant to be used only with manual handling of a IAP client, the usage of hooks was not part of my problem, therefore you will need to debug any issues with the hook yourself, but since hooks are tied to component lifecycles, I think you need to unmount/remove your component for a new instance to be initialized

This is all the info I can give

**Comment by mifi** [comment]
Thanks for the info. I still think it would be nice to keep an issue open for this though, as the problem apparently still exist so it needs to be fixed or documented officially how to properly handle it. Else it's a very sneaky bug that may not appear when testing, but will appear later in production and is hard to reproduce. Shall I create a new issue?

**Comment by mifi** [comment]
Did you by any chance find a way to reproduce the problem locally? I tried:
- turning off wifi
- airplane mode
- force stopping google play services

but `isReadyAndroid` still returns `true`

**Comment by ospfranco** [comment]
I didn't do further testing nor did I upgrade to the latest version as I switched jobs. But we had reverted versions and that worked fine

**Comment by andresesfm** [comment]
It should be fixed in 8.4.0. Please take a look. Reopen if the issue comes back. Thank you
